overlord/snapstate/autorefresh_gating.go

```diff
@@ -32,6 +35,249 @@ import (
 
 var gateAutoRefreshHookName = "gate-auto-refresh"
 
+// cumulative hold time for snaps other than self
+const maxOtherHoldDuration = time.Hour * 48
+
+type holdInfo struct {
+	// FirstHeld keeps the time when the given snap was first held for refresh by a gating snap.
+	FirstHeld time.Time
+	// HoldUntil stores the desired end time for holding.
+	HoldUntil time.Time
+}
+
+var timeNow = func() time.Time {
+	return time.Now()
+}
+
+func lastRefreshed(st *state.State, snapName string) (*time.Time, error) {
+	var snapst SnapState
+	if err := Get(st, snapName, &snapst); err != nil {
+		return nil, fmt.Errorf("internal error, cannot get snap %q: %v", snapName, err)
+	}
+	if snapst.LastRefresh != nil {
+		return snapst.LastRefresh, nil
+	}
+	inf, err := snapst.CurrentInfo()
+	if err != nil {
+		return nil, err
+	}
+	fst, err := os.Stat(inf.MountFile())
+	if err != nil {
+		return nil, err
+	}
+	t := fst.ModTime()
+	return &t, nil
+}
+
+func refreshGating(st *state.State) (map[string]map[string]*holdInfo, error) {
+	// affecting snap -> gating snap(s) -> first-held/hold-until time
+	var gating map[string]map[string]*holdInfo
+	err := st.Get("refresh-gating", &gating)
+	if err != nil && err != state.ErrNoState {
+		return nil, fmt.Errorf("internal error: cannot get refresh-gating: %v", err)
+	}
+	if err == state.ErrNoState {
+		return make(map[string]map[string]*holdInfo), nil
+	}
+	return gating, nil
+}
+
+// HoldError contains the details of snaps that cannot to be held.
+type HoldError struct {
+	SnapsInError map[string]error
+}
+
+func (h *HoldError) Error() string {
+	l := []string{""}
+	for _, e := range h.SnapsInError {
+		l = append(l, e.Error())
+	}
+	return fmt.Sprintf("cannot hold some snaps:%s", strings.Join(l, "\n - "))
+}
+
+// HoldRefresh marks affectingSnaps as held for refresh for up to holdTime.
+// HoldTime of zero denotes maximum allowed hold time.
+// Holding may fail for only some snaps in which case HoldError is returned and
+// it contains the details of failed ones.
+func HoldRefresh(st *state.State, gatingSnap string, holdTime time.Time, affectingSnaps ...string) error {
```

Comment:
holdTime should be a duration, not a time,  yes it's different that how refresh.hold works but that because refresh.hold is mixing api and implementation, people always asked for a duration if possible